Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp

```diff
@@ -218,4 +219,536 @@ Optional<ARIA::Role> HTMLImageElement::default_role() const
+    image_request.set_image_data(image_data);
+
+    Optional<String> crossorigin_attribute;
+    if (auto crossorigin_value = attribute(HTML::AttributeNames::crossorigin); !crossorigin_value.is_null())
+        crossorigin_attribute = String::from_deprecated_string(crossorigin_value).release_value_but_fixme_should_propagate_errors();
+    ListOfAvailableImages::Key key;
+    key.url = url_string;
+    key.mode = cors_setting_attribute_from_keyword(crossorigin_attribute);
+    key.origin = document().origin();
+
+    // 1. If image request is the pending request, abort the image request for the current request,
+    //    upgrade the pending request to the current request
+    //    and prepare image request for presentation given the img element.
+    if (image_request == m_pending_request) {
+        m_current_request->abort(realm());
+        upgrade_pending_request_to_current_request();
+        image_request.prepare_for_presentation(*this);
+    }
+
+    // 2. Set image request to the completely available state.
+    image_request.set_state(ImageRequest::State::CompletelyAvailable);
+
+    // 3. Add the image to the list of available images using the key key, with the ignore higher-layer caching flag set.
+    document().list_of_available_images().add(key, image_data, true).release_value_but_fixme_should_propagate_errors();
+
+    // 4. Fire an event named load at the img element.
+    dispatch_event(DOM::Event::create(realm(), HTML::EventNames::load).release_value_but_fixme_should_propagate_errors());
+
+    set_needs_style_update(true);
+    document().set_needs_layout();
+
+    if (image_data->is_animated() && image_data->frame_count() > 1) {
+        m_current_frame_index = 0;
+        m_animation_timer->set_interval(image_data->frame_duration(0));
+        m_animation_timer->start();
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/images.html#upgrade-the-pending-request-to-the-current-request
+void HTMLImageElement::upgrade_pending_request_to_current_request()
+{
+    // 1. Let the img element's current request be the pending request.
+    VERIFY(m_pending_request);
+    m_current_request = m_pending_request;
+
+    // 2. Let the img element's pending request be null.
+    m_pending_request = nullptr;
+}
+
+void HTMLImageElement::handle_failed_fetch()
+{
+    // AD-HOC
+    dispatch_event(DOM::Event::create(realm(), HTML::EventNames::error).release_value_but_fixme_should_propagate_errors());
+}
+
+// https://html.spec.whatwg.org/multipage/rendering.html#restart-the-animation
+void HTMLImageElement::restart_the_animation()
+{
+    m_current_frame_index = 0;
+    m_animation_timer->start();
+}
+
+// https://html.spec.whatwg.org/multipage/images.html#update-the-source-set
+void update_the_source_set(DOM::Element& element);
+void update_the_source_set(DOM::Element& element)
```

Comment:
We can just make this `static` to avoid the "no previous declaration" error (and also makes sure this method has internal linkage)